src/v/v8_engine/executor.cc

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "v8_engine/executor.h"
+
+#include "seastarx.h"
+
+#include <seastar/core/alien.hh>
+#include <seastar/core/do_with.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/core/smp.hh>
+#include <seastar/util/later.hh>
+
+#include <memory>
+
+namespace v8_engine {
+
+namespace internal {
+
+// condition
+
+condition::condition() noexcept
+  : _file_desc{ss::file_desc::eventfd(0, 0)}
+  , _fd(_file_desc.get())
+  , _pollable_fd{std::move(_file_desc)} {}
+
+ss::future<> condition::wait() {
+    return _pollable_fd
+      .read_some(reinterpret_cast<char*>(&_event), sizeof(_event))
+      .discard_result();
+}
+
+void condition::notify() { ::eventfd_write(_fd, 1); }
+
+// blocking_queue
+
+ss::future<> blocking_queue::start() { return _mutex_for_core.start(1); }
+
+ss::future<> blocking_queue::close() {
+    is_stopped = true;
+    _has_element_cv.notify_all();
+    return _mutex_for_core.invoke_on_all(
+      [](ss::semaphore& sem) { sem.broken(ss::gate_closed_exception()); });
+}
+
+ss::future<> blocking_queue::stop() { return _mutex_for_core.stop(); }
+
+ss::future<> blocking_queue::push(work_item* item) {
+    return _mutex_for_core.local().wait().then([this, item] {
+        if (is_stopped) {
+            throw ss::gate_closed_exception();
+        }
+
+        std::lock_guard lock(_mutex);
```

Comment:
Locking inside the reactor thread doesn't feels right. Why not lock-free queue?

Reply:
fix